src/catalog/src/durable.rs

```diff
@@ -311,6 +314,100 @@ pub trait DurableCatalogState: ReadOnlyDurableCatalogState {
         let id = id.into_element();
         Ok(ReplicaId::User(id))
     }
+
+    /// Migrates storage stash data from `other` to `self`.
+    async fn migrate_storage_state_from(
+        &mut self,
+        other: &mut Box<dyn DurableCatalogState>,
+    ) -> Result<(), CatalogError> {
+        use std::collections::BTreeMap;
+
+        use mz_proto::RustType;
+        use mz_storage_client::controller::StorageTxn;
+
+        let tombstoned = other.get_tombstone().await?;
+        if tombstoned == Some(true) {
+            return Ok(());
+        }
+
+        let mut storage_migration_txn = self.transaction().await?;
+
+        // If the persist-backed implementation of the stash has storage
+        // metadata, this migration already occurred but managed to fail before
+        // committing the stash tombstone.
+        if !storage_migration_txn.get_storage_metadata().is_empty() {
+            let mut tombstone_tx = other.transaction().await?;
+            tombstone_tx.set_tombstone(true)?;
+            return tombstone_tx.commit().await;
+        }
```

Comment:
I don't think you actually even need a tombstone, it's only really useful if you want the ability to rollback. With your current approach you can just always check if `storage_migration_txn.get_storage_metadata().is_empty()` or not since the migration is only one way.